File: cdm/cpp/properties/SEScalarMassPerMass.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "stdafx.h"
#include "properties/SEScalarMassPerMass.h"

const MassPerMassUnit MassPerMassUnit::ug_Per_kg("ug/kg");
const MassPerMassUnit MassPerMassUnit::mg_Per_g("mg/g");

bool MassPerMassUnit::IsValidUnit(const std::string& unit)
{
  if (ug_Per_kg.GetString().compare(unit) == 0)
    return true;
  if (mg_Per_g.GetString().compare(unit) == 0)
    return true;
  return false;
}

const MassPerMassUnit& MassPerMassUnit::GetCompoundUnit(const std::string& unit)
{
  if (ug_Per_kg.GetString().compare(unit) == 0)
    return ug_Per_kg;
  if (mg_Per_g.GetString().compare(unit) == 0)
    return mg_Per_g;
  std::stringstream err;
  err << unit << " is not a valid MassPerMass unit";
  throw CommonDataModelException(err.str());
}

void SEScalarMassPerMass::Load(const cdm::ScalarMassPerMassData& src, SEScalarMassPerMass& dst)
{
  SEScalarMassPerMass::Serialize(src, dst);
}
void SEScalarMassPerMass::Serialize(const cdm::ScalarMassPerMassData& src, SEScalarMassPerMass& dst)
{
  SEScalarQuantity<MassPerMassUnit>::Serialize(src.scalarmasspermass(), dst);
}

cdm::ScalarMassPerMassData* SEScalarMassPerMass::Unload(const SEScalarMassPerMass& src)
{
  if (!src.IsValid())
    return nullptr;
  cdm::ScalarMassPerMassData* dst = new cdm::ScalarMassPerMassData();
  Serialize(src, *dst);
  return dst;
}
void SEScalarMassPerMass::Serialize(const SEScalarMassPerMass& src, cdm::ScalarMassPerMassData& dst)
{
  SEScalarQuantity<MassPerMassUnit>::Serialize(src, *dst.mutable_scalarmasspermass());
}

File: cdm/java/mil/tatrc/physiology/datamodel/properties/SEScalarMassPerMass.java
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

package mil.tatrc.physiology.datamodel.properties;

import com.kitware.physiology.cdm.Properties.ScalarMassPerMassData;

import mil.tatrc.physiology.datamodel.properties.CommonUnits.MassPerMassUnit;

/**
 * @author abray
 * MassPerMass Scalar Class, will enforce that units are proper to MassPerMasss
 */

public class SEScalarMassPerMass extends SEScalar
{
  public SEScalarMassPerMass()
  {
    super();
  }
  
  /**
   * @param value
   * @param unit - specific any valid standard unit abbreviation
   * http://www.bipm.org/en/si/ for this quantity type
   */
  public SEScalarMassPerMass(double value, String unit)
  {
    this();
    this.setValue(value, unit);
  }
  
  /**
   * @param value
   * @param unit - enumeration of commonly used units for this type
   * Units are not limited to this set.
   * You can use the other method convention:(double,string)
   * and specify any valid standard unit abbreviation sting for this type
   * Unit Abbreviation Standards: http://www.bipm.org/en/si/
   */
  public SEScalarMassPerMass(double value, MassPerMassUnit unit)
  {
    this();
    this.setValue(value,unit);
  }
  
  public static void load(ScalarMassPerMassData src, SEScalarMassPerMass dst)
  {
    SEScalar.load(src.getScalarMassPerMass(),dst);
  }
  public static ScalarMassPerMassData unload(SEScalarMassPerMass src)
  {
    if(!src.isValid())
      return null;
    ScalarMassPerMassData.Builder dst = ScalarMassPerMassData.newBuilder();
    unload(src,dst);
    return dst.build();
  }
  protected static void unload(SEScalarMassPerMass src, ScalarMassPerMassData.Builder dst)
  {
    SEScalar.unload(src,dst.getScalarMassPerMassBuilder());
  }
  
  /**
   * @param value
   * @param unit - enumeration of commonly used units for this type
   * Units are not limited to this set.
   * You can use the other method convention:(double,string)
   * and specify any valid standard unit abbreviation sting for this type
   * Unit Abbreviation Standards: http://www.bipm.org/en/si/
   */
  public void setValue(double value, MassPerMassUnit unit)
  {
    this.setValue(value,unit.toString());
  }
  
  /**
   * @param value
   * @param unit - enumeration of commonly used units for this type
   * Units are not limited to this set.
   * You can use the other method convention:(double,string)
   * and specify any valid standard unit abbreviation sting for this type
   * Unit Abbreviation Standards: http://www.bipm.org/en/si/
   */
  public double getValue(MassPerMassUnit unit)
  {
    return this.getValue(unit.toString());
  }

  public boolean validUnit(String unit)
  {
    if(MassPerMassUnit.validUnit(unit))
      return true;
    return false;
  }
}
